trip-cost-calculator-ui/app/components/input.component.tsx

```diff
@@ -1,44 +1,106 @@
 import { useStateWithOverride } from "~/hooks/use-state-with-override.hook";
+import AutosizeInput, { AutosizeInputElement } from "./input/autosize-input";
+import { useLayoutEffect, useRef } from "react";
 
 export type InputProps<T> = {
   value: T;
   onChange: (newValue: T) => void;
   isEditable: boolean;
+  valueFormatter?: (value: T) => string;
   className?: string;
 };
 
 export default function Input<T extends string | number>({
   value: valueProp,
   onChange,
   isEditable,
+  valueFormatter,
   className,
 }: InputProps<T>) {
-  const [value, setValue] = useStateWithOverride(valueProp.toString());
+  const inputRef = useRef<AutosizeInputElement | null>(null);
+
+  const formattedValueProp = valueFormatter
+    ? valueFormatter(valueProp)
+    : valueProp.toString();
+  const [value, setValue] = useStateWithOverride(formattedValueProp);
 
   const isNumberType = typeof valueProp === "number";
-  const inputType = isNumberType ? "number" : "text";
+
+  const cursorPosition = useRef<CursorPosition>({
+    start: 0,
+    end: null,
+    automaticallyAddedChars: 0,
+  });
 
   const changeHandler = (event: React.ChangeEvent<HTMLInputElement>) => {
-    const rawValue = event.target.value;
-    setValue(rawValue);
+    let rawValue = event.currentTarget.value;
+    let coercedValue = rawValue as T;
+    if (isNumberType) {
+      rawValue = rawValue.replaceAll(/[^0-9\.]/g, "");
+
+      let parsedValue = parseFloat(rawValue);
+      if (isNaN(parsedValue)) {
+        parsedValue = 0;
+      }
+      coercedValue = parsedValue as T;
+    }
+
+    if (valueFormatter) {
+      rawValue = valueFormatter(coercedValue);
+    }
+
+    let automaticallyAddedChars = rawValue.length - value.length;
+    if (automaticallyAddedChars > 0) {
+      automaticallyAddedChars -= 1;
+    } else {
+      automaticallyAddedChars += 1;
+    }
+
+    cursorPosition.current = {
+      start: event.currentTarget.selectionStart,
+      end: event.currentTarget.selectionEnd,
+      automaticallyAddedChars,
+    };
 
-    const coercedValue = (isNumberType ? parseFloat(rawValue) : rawValue) as T;
+    setValue(rawValue);
     onChange(coercedValue);
   };
 
+  useLayoutEffect(() => {
+    // Any time we format a value the cursor will get jumped
```

Comment:
This should be testable

Reply:
It proved to be more difficult than expected to get access to the cursor's position from within the test